### ISSUE TITLE
System.InvalidCastException when calling ToQueryString

### ISSUE THREAD

**Issue by yv989c** [dotnet/efcore#30515]
The ADO.NET SQL Client allows the use of types implementing `System.IConvertible` as the [`Value`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.data.sqlclient.sqlparameter.value#remarks) of `SqlParameter` parameters. When trying to use the [`ToQueryString`](https://learn.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.entityframeworkqueryableextensions.toquerystring) method under this situation, it throws a `System.InvalidCastException`.

The following code can reproduce the problem:
```c#
class ConvertibleType : IConvertible
{
    public string ToString(IFormatProvider? provider) => "Hi bug!";
    // Other members omitted for brevity...
}
```
```c#
// db is a valid instance of a DbContext using the SQL Server DB provider.

// First, lest try with a String as the value for the parameter.
var goodQuery = db.Database.SqlQueryRaw<string>(
    "SELECT @p1 AS Value",
    new Microsoft.Data.SqlClient.SqlParameter("@p1", System.Data.SqlDbType.VarChar, -1) { Value = "Hi!" }
    );

var goodQueryResult = goodQuery.Single();
// Result: Hi!

var goodQuerySql = goodQuery.ToQueryString();
// Result:
// DECLARE @p1 varchar(max) = 'Hi!';
// SELECT @p1 AS Value

// Now, lest try with an instance of a type that implements System.IConvertible as the parameter value.
var bugQuery = db.Database.SqlQueryRaw<string>(
    "SELECT @p1 AS Value",
    new Microsoft.Data.SqlClient.SqlParameter("@p1", System.Data.SqlDbType.VarChar, -1) { Value = new ConvertibleType() }
    );

var bugQueryResult = bugQuery.Single();
// Result: Hi bug!

// Will throw System.InvalidCastException.
var bugQuerySql = bugQuery.ToQueryString();
```

### Stack Trace
Using `Microsoft.EntityFrameworkCore.SqlServer v7.0.0`.
```
System.InvalidCastException : Unable to cast object of type 'ConvertibleType' to type 'System.String'.
SqlServerStringTypeMapping.GenerateNonNullSqlLiteral(Object value)
RelationalTypeMapping.GenerateProviderValueSqlLiteral(Object value)
RelationalTypeMapping.GenerateSqlLiteral(Object value)
SqlServerQueryStringFactory.Create(DbCommand command)
FromSqlQueryingEnumerable`1.ToQueryString()
EntityFrameworkQueryableExtensions.ToQueryString(IQueryable source)
```

### Include provider and version information

EF Core version: 5, 6, 7
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET 6.0
Operating system: Windows 11
IDE: Visual Studio 2022 17.5.2

**Comment by roji** [comment]
Not everything that's supported in the underlying ADO.NET driver is necessarily supported as-is by EF - EF is a layer on top which has its own mapping and conversion concepts.

What you're looking for can be done with EF's [value converter feature](https://learn.microsoft.com/en-us/ef/core/modeling/value-conversions?tabs=data-annotations).

**Comment by yv989c** [comment]
Thanks @roji , but I don't see how the EF value converter feature fits in this case. The EF queries are functional as demoed in the provided example.

The issue I'm reporting is about the `ToQueryString` API. There is a problem in the underlying implementation of `ToQueryString` when it tries to project _an idea_ of the underlying SQL, and is making assumptions about the underlying data type contained in the `DbParameter`, causing the exception.

If I'm mistaken about the value converter approach please let me know but I just don't see how they fit here. I'm just passing a parameter that I can use freely in my TSQL expression and not even using entities in the example.

**Comment by roji** [comment]
I looked a bit more into your sample... The problem is that in order to generate a query string (which is a string), EF needs to generate SQL literal representations for the parameters. It does this by using its type mappings, but it doesn't know anything about your special type. In other words, in general EF needs to do more than just put your instances in a SqlParameter.Value - in this case it needs to know how to generate a SQL literal representation of it.

Re value converters, you should be able to use [pre-convention model configuration](https://learn.microsoft.com/en-us/ef/core/modeling/bulk-configuration#pre-convention-configuration) to register a value converter globally for your custom type; the value converter could simply call ToString on it. That should make things work here.

**Comment by yv989c** [comment]
Thanks a lot @roji . I'll give it a shot! 👍

I'll keep you posted.

**Comment by ajcvickers** [comment]
Note from triage: we should at least not throw from ToQueryString.

**Comment by yv989c** [comment]
@roji unfortunately the value converter lead me to a dead end. Under EF 7 I did the following to see if that could resolve the `InvalidCastException` on `ToQueryString`:

Created the following type:
```c#
class ConvertibleTypeValueConverter : ValueConverter<ConvertibleType, string>
{
    public ConvertibleTypeValueConverter()
        : base(v => "", v => new ConvertibleType())
    {
    }
}
```
Then registered it on the relevant `DbContext` per the [documentation](https://learn.microsoft.com/en-us/ef/core/modeling/bulk-configuration#pre-convention-configuration):
```c#
protected override void ConfigureConventions(ModelConfigurationBuilder configurationBuilder)
{
    configurationBuilder
        .Properties<ConvertibleType>()
        .HaveConversion<ConvertibleTypeValueConverter>();
}
```
The error still happens with the same stack trace. I made sure that `ConfigureConventions` was being called by EF in my test, but it doesn't appear to have an effect on what `ToQueryString` does. Of course, I could be doing the registration wrong–I tried swapping the type arguments in `ConvertibleTypeValueConverter` with the same result.

A few observations:
- `ToQueryString` was introduced in EF 5, whereas the `ModelConfigurationBuilder` type was introduced in EF 6.
- Having to override the `DbContext` to register the converter is less than ideal in my scenario.

**Comment by roji** [comment]
> Having to override the DbContext to register the converter is less than ideal in my scenario.

I don't think there's any way in which this could work without something like this; it wouldn't be appropriate for EF to "magically" call ToString() on your type as a fallback because it knows nothing about it. This may make sense for pure string types, but keep in mind that there are other types, and their ToString string representation in .NET may have nothing to do with how you want them to be represented in the database. This is why an explicit mapping guesture is needed via the value converter.

FWIW I think that even at the ADO.NET layer, the SqlClient behavior of automatically converting to strong is dubious, and it certainly wouldn't work in some other database providers if you were to switch. I'd recommend against relying on it in general.

However, if ToQueryString isn't taking into account your explicit HaveConversion, that's probably something we want to look at.

**Comment by ajcvickers** [comment]
@roji This won't work in this case:

```C#
configurationBuilder
        .Properties<ConvertibleType>()
        .HaveConversion<ConvertibleTypeValueConverter>();
```

Because the parameter is not correlated with any property. This might work:

```C#
configurationBuilder
    .DefaultTypeMapping<ConvertibleType>()
    .HasConversion<ConvertibleTypeValueConverter>();
```

But I haven't tried it.

**Comment by yv989c** [comment]
> it wouldn't be appropriate for EF to "magically" call ToString() on your type as a fallback because it knows nothing about it.

Sure, I am not suggesting that. It would have to be more elaborated than that. For example, ADO.NET takes into account the parameter type in order to decide what method to call from `System.IConvertible` (e.g. `SqlDbType.Int` will call `ToInt32`, SqlDbType.Xml will call `ToString`, etc.).

I do understand this is an edge case on this provider, so if there aren't plans to add that complexity at least I would expect `ToQueryString` to just use some placehoder text (e.g. `unknown`) when it encounters something like this instead of the exception that's being thrown by the [following assumption](https://github.com/dotnet/efcore/blob/2ba7dcd19c65e1ed2ab007454de8b67737ee80d0/src/EFCore.SqlServer/Storage/Internal/SqlServerStringTypeMapping.cs#L184).

**Comment by yv989c** [comment]
@ajcvickers, I just tried your suggestion:
```c#
configurationBuilder
    .DefaultTypeMapping<ConvertibleType>()
    .HasConversion<ConvertibleTypeValueConverter>();
```
But it leads to the same outcome.

**Comment by ajcvickers** [comment]
Yeah, I suspect the type mappings are not being used at all in this raw SQL scenario with no mapped types.
